### PR TITLE
Add Custom Attribute De-serialization

### PR DESCRIPTION
**Pull request by richardplatel**
This PR  adds an `attribute` method for providing a custom attribute deserialization class

The `attribute` method takes a field name and a class.  The class must provide a `load` class method that takes 
the attribute value from the file and returns a deserialized version.  ([Similar to ActiveRecord's serialize](https://api.rubyonrails.org/v7.0.4/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html#method-i-serialize))

Deserialization is performed in the code close to where default field values are set, as this seemed like a roughly similar operation to me.

This change removes the need to manually serialize and deserialize, for example, Sorbet T::Enums, when using FrozenRecords

```ruby
class AnimalEnum < T::Enum
  def self.load(value)
    try_deserialize(value)
  end

  enums do
    Dog = new('dog')
    Cat = new('cat')
    Bird = new('bird')
  end
end
```

```ruby
class Price < Object
  attr :value

  class << self
    alias_method :load, :new
  end
  def initialize(value)
    @value = value
  end
end
```

```ruby
class PetLicenseFees < FrozenRecord::Base
  self.base_path = '.'

  attribute :fee, Price
  attribute :animal, AnimalEnum
end
```

```yaml
---
- id: 1
  animal: dog
  fee: 3.50

- id: 2
  animal: cat
  fee: 4.00
```

```ruby
puts (PetLicenseFees.first.attributes)
puts (PetLicenseFees.last.attributes)
```

```
{"id"=>1, "animal"=>#<AnimalEnum::Dog>, "fee"=>#<Price:0x00007fef051b5070 @value=3.5>}
{"id"=>2, "animal"=>#<AnimalEnum::Cat>, "fee"=>#<Price:0x00007fef051ba520 @value=4.0>}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -154,6 +154,28 @@ Composite index keys are not supported.
...
+## Serialized fields
+
+A custom class can be used to deserialize an attribute.  By default `CustomClass.new(value)`, is called where
+`value` is the raw value from the source file.  A method other than `new` can be called by passing a
+`:deserializer_method` name as a symbol.
+
+
+```ruby
+class ContinentString < String ; end
+class CurrencyCodeEnum < T::Enum
```
Comment (casperisfine) [inline README.md:166]: Please don't use `T::` stuff in code sample, that require knowledge of sorbet to understand.

Comment (richardplatel) [reply README.md]: Fixed!

File: spec/frozen_record_spec.rb [hunk 2]
```diff
@@ -74,12 +74,51 @@
       expect(country_model.find_by(name: 'Austria').available).to be == false
     end
 
+
```
Comment (casperisfine) [inline spec/frozen_record_spec.rb:77]:
```suggestion
```

File: lib/frozen_record/base.rb [hunk 3]
```diff
@@ -154,6 +156,13 @@ def add_index(attribute, unique: false)
         self.index_definitions = index_definitions.merge(index.attribute => index).freeze
       end
 
+      def attribute(attribute, klass, deserializer_method: :new)
```
Comment (casperisfine) [inline lib/frozen_record/base.rb:159]: I don't really like that `deserializer_method` thing. I think we should either always use `new` or `load` and if people want to use somthing that doesn't respond to the right interface, they can just do a small facade.

File: lib/frozen_record/base.rb [hunk 4]
```diff
@@ -195,6 +204,9 @@ def load_records(force: false)
 
         @records ||= begin
           records = backend.load(file_path)
+          if attribute_deserializers
+            records = records.map { |r| deserialize_attributes!(r.dup) }
```
Comment (casperisfine) [inline lib/frozen_record/base.rb:208]: It annoys me a bit that we dup the records several times.

We may need to refactor this to avoid the multiple dup, or use a `dup if frozen` or something.

Also you are not re-freezing it here.

File: spec/fixtures/countries.yml.erb [hunk 5]
```diff
@@ -7,8 +7,9 @@
   founded_on: 1867-07-01
   updated_at: 2014-02-24T19:08:06-05:00
   nato: true
-  king: Elisabeth II
+  king: Charles III
```
Comment (casperisfine) [inline spec/fixtures/countries.yml.erb:10]: I appreciate the thought, but I think this PR would be better without this unrelated change.

File: lib/frozen_record/base.rb (hunk 3 above)
Comment (richardplatel) [reply lib/frozen_record/base.rb]: I went with `load` since I think it's likely the attribute classes would also be used with ActiveRecords and ActiveRecord serialization uses `load`.

We could be _very_ clever and do something like `if klass.responds_to(:load) klass.load(value) else klass.new(value)` but I think that would lead to unexpected behaviour for callers.

File: lib/frozen_record/base.rb (hunk 4 above)
Comment (richardplatel) [reply lib/frozen_record/base.rb]: I refactored it so it's at least no worse than it was before (I think).  

I tried to refactor further, getting rid of another `.map` and `.dup` by removing 
```
if attribute_deserializers.any? || default_attributes
  records = records.map { |r| assign_defaults!(deserialize_attributes!(r.dup)).freeze }.freeze
end
```

and changing 
```diff
-  records = FrozenRecord.ignore_max_records_scan { records.map { |r| load(r) }.freeze }
+  records = FrozenRecord.ignore_max_records_scan { records.map { |r| new(r.dup).freeze }.freeze }
```
but that required changes to listing attributes and the change got too big.   I think it could be done but maybe in a separate PR?

File: lib/frozen_record/base.rb [hunk 6]
```diff
@@ -3,6 +3,8 @@
 require 'active_support/descendants_tracker'
 require 'frozen_record/backends'
 
+require 'byebug'
+
```
Comment (casperisfine) [inline lib/frozen_record/base.rb:7]:
```suggestion
```

File: README.md [hunk 7]
```diff
@@ -154,6 +154,41 @@ Composite index keys are not supported.
...
+
+class CurrencyCodeEnum
+  CAD = "CAD"
+  EUR = "EUR"
+  class << self
+    def load(value)
+      if value == "CAD"
+          CAD
+      elsif value == "EUR"
+          EUR
```
Comment (casperisfine) [inline README.md:178]: Nitpick:, this example is a bit weird, because you could just use `String#-@` to the same effect.

Do you have any better example? I don't know some kind of `Struct.new`? Or even a `Money` class?

**Review by casperisfine — APPROVED** [review]
Ok, one last thing after that I'm good.
